packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -158,56 +198,71 @@ function getNodeSchemaType(nodeSchema: TreeNodeStoredSchema): SchemaFactoryNodeK
  * It is up to the caller to determine whether such discrepancies matter.
  */
 export function* getAllowedContentDiscrepancies(
-	view: TreeStoredSchema,
+	view: FieldSchema,
 	stored: TreeStoredSchema,
 ): Iterable<Discrepancy> {
 	// check root schema discrepancies
-	yield* getFieldDiscrepancies(
-		view.rootFieldSchema,
-		stored.rootFieldSchema,
-		undefined,
-		undefined,
-	);
+	yield* getFieldDiscrepancies(view, stored.rootFieldSchema, undefined, undefined);
+
+	const storedAllowedTypes = stored.nodeSchema;
+
+	// collect all annotated allowed types from the view schema
+	const annotatedAllowedTypes: AnnotatedAllowedType<TreeNodeSchema>[] = [];
+	walkFieldSchema(view, {
+		allowedTypes: (allowedTypes) => {
+			annotatedAllowedTypes.push(...allowedTypes.types);
+		},
+	});
+
+	const viewAllowedTypes = new Map<TreeNodeSchemaIdentifier, TreeNodeSchema>();
+	for (const annotatedAllowedType of annotatedAllowedTypes) {
+		const { type } = annotatedAllowedType;
+		// map view schema identifiers to the field schemas to make access in the stored schema pass more efficient
+		const identifier: TreeNodeSchemaIdentifier = brand(type.identifier);
+		viewAllowedTypes.set(identifier, type);
+
+		const storedSchema = stored.nodeSchema.get(identifier);
+
+		// if the view schema has an allowed type that's not in the stored schema
+		if (!storedAllowedTypes.has(identifier) || storedSchema === undefined) {
+			const viewType = getViewNodeSchemaType(type);
+			// TODO does it make sense to have this mismatch when there will also be an allowedTypes mismatch?
+			yield {
+				identifier,
+				mismatch: "nodeKind",
+				view: viewType,
+				stored: undefined,
+			};
+		} else {
+			yield* getNodeDiscrepancies(identifier, annotatedAllowedType, storedSchema);
+		}
+	}
 
-	for (const result of compareMaps(view.nodeSchema, stored.nodeSchema)) {
-		switch (result.type) {
-			case "aExtra": {
-				const viewType = getNodeSchemaType(result.value);
-				yield {
-					identifier: result.key,
-					mismatch: "nodeKind",
-					view: viewType,
-					stored: undefined,
-				};
-				break;
-			}
-			case "bExtra": {
-				const storedType = getNodeSchemaType(result.value);
-				yield {
-					identifier: result.key,
-					mismatch: "nodeKind",
-					view: undefined,
-					stored: storedType,
-				};
-				break;
-			}
-			case "both": {
-				yield* getNodeDiscrepancies(result.key, result.valueA, result.valueB);
-				break;
-			}
-			default:
-				break;
+	for (const identifier of storedAllowedTypes.keys()) {
```

Comment:
rather than iterating over the keys of this map, then looking up the values from the same map below via stored.nodeSchema.get, I think just iterating of the map (and thus getting its key value pairs) would be both more efficient and easier to understand.